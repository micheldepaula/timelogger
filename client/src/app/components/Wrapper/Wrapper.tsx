import * as React from 'react'
import { Provider } from 'react-redux'
import { store } from '../../store'

interface WrapperProps {
  children: JSX.Element | JSX.Element[]
}

const Wrapper: React.FC<WrapperProps> = ({ children }) => {
  return <Provider store={store}>{children}</Provider>
}

export default Wrapper
