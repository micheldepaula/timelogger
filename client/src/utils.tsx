export interface IProject  {
    id?: number ,
    name?: string ,
    hours?: number | string,
    end?: Date | string

}
