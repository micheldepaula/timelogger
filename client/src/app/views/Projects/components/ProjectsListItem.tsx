import { useAppDispatch } from '../../../hooks'
import React, { useState } from 'react'
import { IProject } from '../../../../utils'
import { editProject } from '../../../slice/projectSlice'

type ProjectsListItemProps = {
  project: IProject
}

const ProjectsListItem: React.FC<ProjectsListItemProps> = ({ project }) => {
  const [edit, setEdit] = useState<boolean>(false)
  const [hours, setHours] = useState(project.hours)
  const [name, setName] = useState(project.name)
  const dispatch = useAppDispatch()
  const handleEditProject = () => {
    setEdit(false)
    if (project.end) {
      dispatch(
        editProject({
          id: project.id,
          name: name,
          hours: Number(hours),
          end: new Date(project.end.toString()),
        }),
      )
    }
  }
  const EditableInputName = () => {
    return (
      <td className='border px-4 py-2'>
        <input
          className='border rounded-full py-2 px-4'
          value={name}
          type='name'
          placeholder='Project Name'
          aria-label='Project Name'
          onChange={(e) => setName(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditableInputHours = () => {
    return (
      <td className='border px-4 py-2 '>
        <input
          className='border rounded-full py-2 px-4'
          value={hours}
          type='number'
          placeholder='DeadLine'
          aria-label='DeadLine'
          onChange={(e) => setHours(e.target.value.toString())}
        />
      </td>
    )
  }

  const EditButtons = () => {
    return (
      <div className='flex justify-between '>
        <button
          className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={handleEditProject}
          data-testid={`projects-list-item-${project.id}-button-edit`}
        >
          Save
        </button>
        <button
          className='bg-red-500 hover:bg-red-700 text-white rounded-full py-2 px-4 ml-2'
          onClick={() => setEdit(false)}
          data-testid={`projects-list-item-${project.id}-button-cancel-edit`}
        >
          cancel
        </button>
      </div>
    )
  }

  return (
    <tr data-testid={`projects-list-item-${project.id}`}>
      <td className='border px-4 py-2 w-12'>{project.id}</td>
      {!edit ? <td className='border px-4 py-2'>{name}</td> : <EditableInputName />}
      {!edit ? <td className='border px-4 py-2'>{hours}</td> : <EditableInputHours />}
      <td className='flex justify-between border px-4 py-2 '>
        {project.end && new Date(project.end).toISOString().split('T')[0]}
        {!edit ? (
          <button
            className='bg-blue-500 hover:bg-blue-700 text-white rounded-full py-2 px-4 ml-2'
            onClick={() => setEdit(true)}
          >
            Edit
          </button>
        ) : (
          <EditButtons />
        )}
      </td>
    </tr>
  )
}
export default ProjectsListItem
