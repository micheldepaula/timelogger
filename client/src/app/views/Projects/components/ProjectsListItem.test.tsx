import React from 'react';
import { render } from '@testing-library/react';
import ProjectsListItem from './ProjectsListItem';
import Wrapper from '../../../components/Wrapper/Wrapper';
import { IProject } from '../../../../utils';

jest.mock('react-redux', () => {
  const actualModule = jest.requireActual('react-redux');
  return {
    __esModule: true,
    ...actualModule,
  };
});

describe('Projects List Item Component', () => {
  const projectItem: IProject = { end: '2023-06-29', hours: 8, id: 1, name: 'Project List Item' };

  const rerender = () =>
    render(
      <Wrapper>
        <ProjectsListItem project={projectItem} />
      </Wrapper>,
    );

  it('renders correctly', () => {
    const { asFragment, getByTestId } = rerender();
    expect(getByTestId(`projects-list-item-${projectItem.id}`)).toBeDefined();
    expect(asFragment).toMatchSnapshot();
  });
});
