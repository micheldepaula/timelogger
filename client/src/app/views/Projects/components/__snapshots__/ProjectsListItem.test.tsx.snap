// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects List Item Component renders correctly 1`] = `[Function]`;
